presto-hive/src/main/java/com/facebook/presto/hive/s3/PrestoS3FileSystem.java

```diff
@@ -914,28 +1081,44 @@ private static Optional<AWSCredentials> getAwsCredentials(URI uri, Configuration
         if (isNullOrEmpty(accessKey) || isNullOrEmpty(secretKey)) {
             return Optional.empty();
         }
-        return Optional.of(new BasicAWSCredentials(accessKey, secretKey));
+        return Optional.of(AwsBasicCredentials.create(accessKey, secretKey));
+    }
+
+    /**
+     * This exception is for stopping retries for S3 calls that shouldn't be retried.
+     * For example, "Caused by: com.amazonaws.services.s3.model.AmazonS3Exception: Forbidden (Service: Amazon S3; Status Code: 403 ..."
+     */
+
+    @VisibleForTesting
+    static class UnrecoverableS3OperationException
+            extends IOException
+    {
+        public UnrecoverableS3OperationException(Path path, Throwable cause)
+        {
+            // append the path info to the message
+            super(format("%s (Path: %s)", cause, path), cause);
+        }
     }
 
     public static class PrestoS3ObjectMetadata
     {
-        private final ObjectMetadata objectMetadata;
         /**
          * Certain filesystems treat empty directories as zero byte objects and their name ends with a path separator, i.e. '/'.
          * To fetch ObjectMetadata for such keys, the path separator needs to be appended to the key otherwise null is returned.
          * This field denotes whether a path separator was appended to the key while fetching the metadata for given path.
          */
+        private final S3Response objectResponse;
```

Comment:
nit: let's move this before the above comment. The comment is for the variable `keyNeedsPathSeparator`.